ledger/complete/mtrie/flattener/iterator.go

```diff
@@ -75,6 +84,30 @@ func NewNodeIterator(mTrie *trie.MTrie) *NodeIterator {
 	return i
 }
 
+// NewUniqueNodeIterator returns a node NodeIterator, which iterates through all unique nodes
```

Comment:
Interesting point!  I think recursive DFS can work and recursive calls would increase as trie depth grows.  Currently, we allocate array (stack) once with 256 elements (I can optimize it to be fewer).

My hunch is recursion would probably be slower (based past experience with old compilers and function call overhead).  When time allows, I may do benchmark comparisons with recursion to see how it goes with Go 1.17+.

Reply:
I implemented this algorithm originally. You are right that it is a DFS. The non-recursive stack version is more efficient because it avoids the usage of the stack. (at least assuming that the compiler is not able to turn the recursive into the non-recursive version; at least back then, the recursive version was noticeably slower).

The stack size of 256 is given by the max depth of the tree. We now have the `maxDepth` field (the max number of edges). I think we could use this instead of the full length. Not sure that this would result in any noteworthy performance gain, because for a realistic trie, we need probably anyway around 60 - 100 elements.

Reply:
Thank you both for the details! 
Interesting, I didn't expect the recursive version to be much slower than the non-recursive version. Plus I thought the simplicity of the recursive code is worth the time overhead. It looks like no :) I guess the call stack turns out to be much more expensive to manage than the "manual" stack.

 > When time allows, I may do benchmark comparisons with recursion to see how it goes with Go 1.17+
 No rush for this @fxamacker, I guess you already optimized this process significantly so there is no reason to spend more time on it for now.